src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -59,6 +60,58 @@ namespace ErrorCodes
     extern const int UNFINISHED;
 }
 
+class DatabaseNameHints : public IHints<1, DatabaseNameHints>
+{
+public:
+    explicit DatabaseNameHints(const DatabaseCatalog & database_catalog_)
+        : database_catalog(database_catalog_)
+    {
+    }
+    Names getAllRegisteredNames() const override
+    {
+        Names result;
+        auto databases_list = database_catalog.getDatabases();
+        for (const auto & database_name : databases_list | boost::adaptors::map_keys)
+        {
+            if (database_name == DatabaseCatalog::TEMPORARY_DATABASE)
+                continue;
+            result.emplace_back(database_name);
+        }
+        return result;
+    }
+private:
+    const DatabaseCatalog & database_catalog;
+};
+
+class TableNameHints : public IHints<1, TableNameHints>
+{
+public:
+    TableNameHints(const DatabaseCatalog & database_catalog_, ContextPtr context_, const String & database_name_)
+        : database_catalog(database_catalog_)
+        , context(context_)
+        , database_name(database_name_)
+    {
+    }
+    Names getAllRegisteredNames() const override
+    {
+        Names result;
+        DatabasePtr database = database_catalog.tryGetDatabase(database_name);
+        if (database)
+        {
+            for (auto table_it = database->getTablesIterator(context); table_it->isValid(); table_it->next())
+            {
+                const auto & storage_id = table_it->table()->getStorageID();
+                result.emplace_back(storage_id.getTableName());
+            }
+        }
+        return result;
+    }
+private:
+    const DatabaseCatalog & database_catalog;
+    ContextPtr context;
+    String database_name;
+};
```

Comment:
Let's make it usable in IDatabase context - we need to use it in IDatabase::getTable
First, we need to take it out to dedicated header.
Then let's replace all data members with one `DatabasePtr`, and resolve it in constructor.
Then let's add constructor from `DatabasePtr` also - to use it when we already have a database.

After all of this we need to add hints to the IDatabase::getTable.